common/filesystem.h

```diff
@@ -0,0 +1,1374 @@
+  // parent directory.
+  auto OpenDir(std::filesystem::path path,
+               CreationFlags creation_flags = OpenExisting,
+               ModeType creation_mode = 0700) -> ErrorOr<Dir, PathError>;
+
+  // Reads the file at the provided path to a string.
+  //
+  // This is a convenience wrapper for opening the path, reading the returned
+  // file to a string, and closing it. Errors from any step are returned.
+  auto ReadFileToString(std::filesystem::path path)
+      -> ErrorOr<std::string, PathError>;
+
+  // Writes the provided `content` to the provided path.
+  //
+  // This is a convenience wrapper for opening the path, creating it according
+  // to `creation_flags` as necessary, writing `content` to it, and closing it.
+  // Errors from any step are returned.
+  auto WriteFileFromString(std::filesystem::path path, llvm::StringRef content,
+                           CreationFlags creation_flags = CreateAlways)
+      -> ErrorOr<Success, PathError>;
+
+  // Changes the current working directory to this directory.
+  auto Chdir() -> ErrorOr<Success, FdError>;
+
+  // Changes the current working directory to the provided path.
+  //
+  // An error if the provided path is not a directory. Does not open the
+  // provided path as a directory, but it will be available as the current
+  // working directory via `Cwd()`.
+  auto Chdir(std::filesystem::path path) -> ErrorOr<Success, PathError>;
+
+  // Creates a symlink at the provided path with the contents of `target`.
+  //
+  // Note that the target of a symlink is an arbitrary string and there is no
+  // error checking on whether it exists or is sensible. Also, the target string
+  // set will be up to the first null byte in `target`, regardless of its
+  // `size`. This will not overwrite an existing symlink at the provided path.
+  auto Symlink(std::filesystem::path path, const std::string& target)
+      -> ErrorOr<Success, PathError>;
+
+  // Creates the directories in the provided path, using the permissions in
+  // `creation_mode`.
+  //
+  // This will create any missing directory components in `path`. Relative paths
+  // will be created relative to this directory, and without re-resolving its
+  // path. The leaf created directory is opened and returned.
+  //
+  // The implementation allows for concurrent creation of the same directory (or
+  // a prefix) without error or corruption. However, concurrent creation of the
+  // same path tree cannot be made safe using this API. This function works to
+  // ensure that any directories created have the correct permissions and user,
+  // but only the _first_ concurrent creation will detect the replacement of its
+  // created directory with a potentially untrusted entity. Once some prefix has
+  // been redirected to an untrustworthy location, subsequent calls cannot
+  // detect and avoid this. When creating directories securely, start from a
+  // securely directly created `Dir` with limited access, and then create
```

Comment:
```suggestion
  // directly created `Dir` with limited access, and then create
```
I don't know that "securely" is adding anything except confusion here. We're describing the secure process, so we don't need to ask the user to do extra secure unspecified things